### PR TITLE
Conditionally removing fields from introspection

### PR DESCRIPTION
**Pull request by andcan**
First work on https://github.com/99designs/gqlgen/issues/380.

 - Added new interface for plugins: `SchemaMutator`: I need to be able to mutate schema before code generation.
	Plugins implementing this interface get called between `ConfigMutator.MutateConfig` and `CodeGenerator.GenerateCode`.
	I had to modify `codegen.BuildData` signature and add a couple rows into `api.Generate` to make this work
 - Added `introspection` plugin.
	Injects `@introspection` directive on `__Type.inputFields` and `__Type.fields`.
	Generates `@introspection` directive implementation.
- Added `introspection` example.

This PR is to receive comments/suggestions on how to implement this.
If/when looks good I'll add tests and docs.

### CODE REVIEW

File: plugin/introspection/plugin.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	return &Plugin{cfg: cfg}
+}
+
+func (p *Plugin) Name() string {
+	return "introspection"
+}
+
+func (p *Plugin) MutateConfig(cfg *config.Config) error {
+	for _, d := range p.cfg.Directives {
+		cfg.Directives[d] = config.DirectiveConfig{SkipRuntime: true}
```
Comment (vvakame) [inline plugin/introspection/plugin.go]: I'm already using `@hasRole` directive, and it has runtime implementation.
I think `MutateConfig` is not necessary. 
user defines `@hide` and `@requireAuth` in schema. `skip_runtime` settings should write gqlgen.yml explicitly.

File: plugin/introspection/plugin.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	introspection := &ast.DirectiveDefinition{
+		Description: "",
+		Name:        "introspection",
+		Arguments:   nil,
+		Locations: []ast.DirectiveLocation{
+			ast.LocationInputFieldDefinition,
+			ast.LocationFieldDefinition,
+		},
+		Position: nil,
+	}
```
Comment (vvakame) [inline plugin/introspection/plugin.go]: `@introspection` does not have an any (common) specs. I want to hide it in IntrospectionQuery.
I recommend add something like `SkipRuntime: true` or `Builtin: true` .
but wanna generate directives in code. 🤔 

@vektah do you have any opinion in this topic?

File: plugin/introspection/plugin.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
+package introspection
```
[the remark this answers is not in the document]
Comment (andcan) [reply plugin/introspection/plugin.go]: Yes, you're right. I renamed it to introauth, could't find a better naeìme.

File: plugin/introspection/plugin.go (hunk 1 above)
Comment (andcan) [reply plugin/introspection/plugin.go]: Yes you're right, this is not necessary.

File: codegen/data.go [hunk 4]
```diff
@@ -35,17 +35,14 @@ type builder struct {
 	Directives map[string]*Directive
 }
 
-func BuildData(cfg *config.Config) (*Data, error) {
+func BuildData(cfg *config.Config, schema *ast.Schema, schemaStr map[string]string) (*Data, error) {
```
[the remark this answers is not in the document]
Comment (vektah) [reply codegen/data.go]: I think I like this implementation better https://github.com/99designs/gqlgen/pull/851/files#diff-86f7257ff3b7e8564cb7fae47bd53676R37-R46

Do we actually need to modify schema or can it be purely additive?

File: plugin/introspection/plugin.go (hunk 2 above)
Comment (vektah) [reply plugin/introspection/plugin.go]: Yep, it already exists https://github.com/99designs/gqlgen/blob/master/codegen/directive.go#L24

**Comment by vektah** [comment]
The [handler refactor](https://github.com/99designs/gqlgen/pull/885) pr has some overlap with this, between the two we should be able to pull introspection out entirely into a [handler extension](https://github.com/99designs/gqlgen/pull/885/files#diff-a2eafa7f07f95598428b68504f03e5ebR1-R18).

Might be good to sort out what that looks like before we cut a release, so we get the interfaces for generated handler plugins right.

**Comment by vvakame** [comment]
@andcan #885 is merged. could you merge master/HEAD into this PR?
I really need this changes 😉 please let me know I can help you anything.

**Comment by andcan** [comment]
Sorry lost that, I will give a look to the handler refactor but I'm a bit out of time right now it might take some time. I'll let you know if I need help.

**Comment by vvakame** [comment]
@andcan 👀

**Comment by vektah** [comment]
Closing this until its rebased and ready for review again, feel free to reopen.

**Pull request closed without merging**